pkg/pool-weighted/contracts/WeightedMath.sol

```diff
@@ -407,63 +407,58 @@ library WeightedMath {
     ) internal pure returns (uint256 protocolFeeAmount) {
+        uint256 supplyGrowthRatio = currentSupply.divDown(previousSupply);
+
         // Joins and exits are symmetrical; for simplicity, we consider a join, where the invariant and supply
         // both increase.
 
-        // |-------------------------|-- postJoinExitInvariant
+        // |-------------------------|-- original invariant * invariantGrowthRatio
         // |   increase from fees    |
         // |-------------------------|-- original invariant * supply growth ratio (fee-less invariant)
         // |                         |
         // | increase from balances  |
-        // |-------------------------|-- preJoinExitInvariant
+        // |-------------------------|-- original invariant
         // |                         |
-        // |                         |  |------------------|-- postJoinExitSupply
+        // |                         |  |------------------|-- currentSupply
         // |                         |  |    BPT minted    |
-        // |                         |  |------------------|-- preJoinExitSupply
+        // |                         |  |------------------|-- previousSupply
         // |   original invariant    |  |  original supply |
         // |_________________________|  |__________________|
         //
         // If the join is proportional, the invariant and supply will likewise increase proportionally,
-        // so the growth ratios (postJoinExit / preJoinExit) will be equal. In this case, we do not charge
+        // so the growth ratios (invariantGrowthRatio / supplyGrowthRatio) will be equal. In this case, we do not charge
         // any protocol fees.
-        //
+
+        if (supplyGrowthRatio == invariantGrowthRatio || protocolSwapFeePercentage == 0) return 0;
+
         // If the join is non-proportional, the supply increase will be proportionally less than the invariant increase,
         // since the BPT minted will be based on fewer tokens (because swap fees are not included). So the supply growth
         // is due entirely to the balance changes, while the invariant growth also includes swap fees.
         //
         // To isolate the amount of increase by fees then, we multiply the original invariant by the supply growth
         // ratio to get the "feeless invariant". The difference between the final invariant and this value is then
         // the amount of the invariant due to fees, which we convert to a percentage by normalizing against the
-        // final (postJoinExit) invariant.
+        // final invariant.
         //
-        // Compute the portion of the invariant increase due to fees
-        uint256 supplyGrowthRatio = postJoinExitSupply.divDown(preJoinExitSupply);
+        // This operation can be simplified to give the below expression for fraction of the Pool which is swap fees.
+        // This allows us to work in terms of ratios of invariants rather than true values.
```

Comment:
// This operation can be simplified to the expression given below for the fraction of the Pool which is swap fees.
// This allows us to work in terms of ratios of invariants rather than absolute values.